apps/server/src/state.ts

```diff
@@ -0,0 +1,456 @@
+                state.timer.timerType,
+              );
+            }
+
+            state.timer.current = getCurrent(
+              state.timer.startedAt,
+              state.timer.duration,
+              state.timer.addedTime,
+              state.timer.pausedTime,
+              state.timer.clock,
+              state.timer.timeEnd,
+              state.timer.timerType,
+            );
+
+            state.timer.elapsed = state.timer.duration - state.timer.current;
+
+            return { isFinished };
+          }
+
+          let _force = force;
+          let _didUpdate = false;
+          let _doRoll = false;
+          let _isFinished = false;
+          let _shouldNotify = false;
+
+          const previousTime = state.timer.clock;
+          state.timer.clock = clock.timeNow();
+
+          if (previousTime > state.timer.clock) {
+            _force = true;
+          }
+
+          // we call integrations if we update timers
+          if (state.playback === Playback.Roll) {
+            const result = roll();
+            _shouldNotify = true;
+            _doRoll = result.doRoll;
+            _isFinished = result.isFinished;
+          } else if (state.timer.startedAt !== null) {
+            // we only update timer if a timer has been started
+            const result = play();
+            _shouldNotify = true;
+            _isFinished = result.isFinished;
+          }
+
+          // we only update the store at the updateInterval
+          // side effects such as onFinish will still be triggered in the update functions
+          if (_force || state.timer.clock > state.timer.lastUpdate + updateInterval) {
+            state.timer.lastUpdate = state.timer.clock;
+            _didUpdate = true;
+          }
+
+          return {
+            didUpdate: _didUpdate,
+            doRoll: _doRoll,
+            isFinished: _isFinished,
+            shouldNotify: _shouldNotify,
+          };
+        },
+        {
+          sideEffect({ didUpdate, doRoll, isFinished, shouldNotify }, newState) {
+            if (didUpdate && shouldNotify) {
+              integrationService.dispatch(TimerLifeCycle.onUpdate);
+            }
+
+            if (doRoll) {
+              PlaybackService.roll();
+            }
+
+            if (isFinished) {
+              integrationService.dispatch(TimerLifeCycle.onFinish);
+              if (newState.playback === Playback.Play) {
+                if (newState.timer.endAction === EndAction.Stop) {
+                  PlaybackService.stop();
+                } else if (newState.timer.endAction === EndAction.LoadNext) {
+                  // we need to delay here to put this action in the queue stack. otherwise it won't be executed properly
+                  setTimeout(() => {
+                    PlaybackService.loadNext();
+                  }, 0);
+                } else if (newState.timer.endAction === EndAction.PlayNext) {
+                  PlaybackService.startNext();
```

Comment:
it feels strange to loop back to the parent service. 

the order of execution is
Controller > PlaybackService > TimerService > state

How would we return a result from the evaluation that is not in state (as in isFinished?) Maybe this needs to be calculated at the parent?
is there a more elegant way to escalate this information up? 

Perhaps we deprecate the entire `TimerService` in favour of the `PlaybackService`?